src/inmanta/ast/statements/generator.py

```diff
@@ -181,9 +181,14 @@ def __init__(self, stmt: "For", resolver: Resolver, queue: QueueScheduler) -> No
         self.stmt = stmt
         self.seen: set[int] = set()
 
-    def receive_result(self, value: object, location: ResultVariable) -> None:
+    @classmethod
+    def pure_gradual(self) -> bool:
+        # freezing the iterable allows dropping eager promises
+        return False
```

Comment:
but it will never cause an additional instance to be added to a relation directly. 

I find this one very weird. 

For loops are the most gradual thing there is, they are pure for each, they only produce something if there is more ouput. 


I would expect this to break many models, because the first relation feeding into a for loop will possible get progress potential. But the one feeding into it does not.

Reply:
Added a test case that triggers what you describe in 1200114.